datadog/resource_datadog_organization.go

```diff
@@ -0,0 +1,357 @@
+package datadog
+
+import (
+	"context"
+
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"
+
+	datadogV1 "github.com/DataDog/datadog-api-client-go/api/v1/datadog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func resourceDatadogOrganization() *schema.Resource {
```

Comment:
Given the current limitations of the organizations API, that prevent you from interacting with any child organization from the parent one, I'm hesitant to add a resource for it at all. You wouldn't be able to do much via terraform with this.
I would keep the organization settings resource only, this one makes sense.
What do you think ?

Reply:
In our case we want to dynamically create DataDog organizations which can later be configured using by organization settings resource.
Also to create child organization we need to use "root" account API and APP keys (and use one terraform provider) and to adjust organization settings we need use use child organization API and APP keys (another terraform provider). If we want to do both, create and change organization settings, in one resource, we would need to provide two sets of credentials to one resource.

Reply:
Well that's my point exactly, I wouldn't add org creation to terraform at all. I'd just keep the org settings resource as it is currently in your PR, and remove the child org one.
Because with the child org resource, you'd only be able to create, then you can't even read, update or delete, so I don't think it makes sense for terraform.

Reply:
OK. So let's discuss a few cases.
1. One resource to create and manage settings of the organization.

```hcl
provider "datadog" {
  api_key = "root"
  app_key = "root"
}

resource "datadog_organization" "organization" {
  name = "someName"
}
```

In this case, we will be able to create the organization using root account credentials. If we want to change something, for example, the name of the organization, we would need to use child organization API and application keys. Thankfully, those keys are returned by the POST https://api.datadoghq.com/api/v1/org request. But if the API or application key from the child organization will be removed we will lose possibilities to manage child organization. As a workaround we could provide child organization credentials as parameters:
```hcl
provider "datadog" {
  api_key = "root"
  app_key = "root"
}

resource "datadog_organization" "organization" {
  name = "someName"

  api_key = "child"
  app_key = "child"
}
```

This looks odd. Secondly, API and application keys are stored in the tfstate file.

2. Another problem. As I know, organization API is not available for everyone. If the team has only one organization, how they should manage that organization through Terraform? Since there is only one organization they probably will end up in something like that:

```hcl
provider "datadog" {
  api_key = "org"
  app_key = "org"
}

resource "datadog_organization" "organization" {
  name = "someName"

  api_key = "org"
  app_key = "org"
}
```

In addition, they would need to import `datadog_organization` resource before executing `plan` or `apply`.

3. If we split organization creation and organization settings (basically mirror what is possible through the API), we will end up with this:

```hcl
provider "datadog" {
  alias   = "root"
  api_key = "root"
  app_key = "root"
}

resource "datadog_organization" "organization" {
  name = "someName"

  lifecycle {
    ignore_changes = [name]
  }

  provider = datadog.root
}

provider "datadog" {
  api_key  = datadog_organization.organization.api_key.0.key
  app_key  = datadog_organization.organization.application_key.0.hash
  validate = false
}

resource "datadog_organization_settings" "organization" {
  name = "someName"
}
```

It is a lot, I agree, but we are flexible. If the API or application keys will be removed from the organization, we can create a new one and pass them to the child provider, we will not pass them to the resource like in example 1 or 2. Secondly, if the team has only one organization and they are unable to create child organization, they can just use `datadog_organization_settings` like:

```hcl
provider "datadog" {
  api_key  = "child"
  app_key  = "child"
}

resource "datadog_organization_settings" "organization" {
  name = "someName"
}
```

I hope, it is clear tight now. By having two different resources, we can create organizations and manage them. In our case, we are more focus on organization creation, then on managing organizations. We want to create organization per product which we are creating, and we have a lot of products. So having a resource to create the org is something crucial for us.

Reply:
So yeah I totally agree with you, it is better to have a dedicated resource for managing an org, separated from org creation. As you mentioned, it also makes sense considering not everybody has access to the child org API.
So your `datadog_organization_settings` resource looks good to me as it is.

My problem was with the `datadog_organization` resource itself, as it is very limited in functionality, you can only create the org, and then you could basically remove it from the terraform configuration and it wouldn't really matter.
But your point makes sense, let's keep it.